### PR TITLE
feat: [HL-33] Reusable loading view on button ♻️

### PR DESCRIPTION
**Pull request by Ahattalla**
#### Proposed Changes
I created UILoadingButton that shows loading indicator on it

#### Screenshot
Screenshot 2023-06-20 at 3 52 48 AM
Screenshot 2023-06-20 at 3 53 14 AM

#### Checklist

- [x] I have tested these changes locally.
- [ ] I have added appropriate documentation or updated existing documentation.
- [ ] I have added appropriate test coverage or updated existing test coverage.
- [ ] I have updated the changelog (if applicable).
- [x] I have followed the project's code style and formatting guidelines.
- [x] I have reviewed and adhered to the project's contributing guidelines.

### CODE REVIEW

File: Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self.addConstraint(xCenterConstraint)
+
+        let yCenterConstraint = NSLayoutConstraint(item: self,
+                                                   attribute: .centerY,
+                                                   relatedBy: .equal,
+                                                   toItem: activityIndicator,
+                                                   attribute: .centerY,
+                                                   multiplier: 1,
+                                                   constant: 0)
+        self.addConstraint(yCenterConstraint)
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift]:
```suggestion
        guard let activityIndicator else {
            assertionFailure("Trying to configure constraints for invalid activity indicator.")
            return
        }
        
        NSLayoutConstraint.activate([
            activityIndicator.centerXAnchor.constraint(equalTo: centerXAnchor),
            activityIndicator.centerYAnchor.constraint(equalTo: centerYAnchor),
        ])
    }
```

File: Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+}
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift]:
```suggestion
    }
}
```

File: Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import UIKit
+
+class UILoadingButton: UIButton, AnimatableView {
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift]:
```suggestion
final class UILoadingButton: UIButton, AnimatableView {
```

File: Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    var originalButtonText: String?
+    var originalImage: UIImage?
+    var activityIndicator: UIActivityIndicatorView?
+
+    @IBInspectable var activityIndicatorColor: UIColor = .white
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/UILoadingButton/UILoadingButton.swift]:
```suggestion
    var activityIndicatorColor: UIColor = .white
    
    private var originalButtonText: String?
    private var originalImage: UIImage?
    private var activityIndicator: UIActivityIndicatorView?
```

File: Healthy/Classes/ReusableViews/UILoadingButton/AnimatableView.swift [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import Foundation
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/UILoadingButton/AnimatableView.swift]: Thanks for adding abstraction...

**Review by ahmdmhasn — APPROVED** [review]
While this is good, but if we already have a subclassed button and we need to apply this animation, it would be very hard. 

I would suggest you to reach @oradwanomar for an idea about this and let's discuss if needed. 

Great effort. 🎉

**Comment by oradwanomar** [comment]
> While this is good, but if we already have a subclassed button and we need to apply this animation, it would be very hard.
> 
> I would suggest you to reach @oradwanomar for an idea about this and let's discuss if needed.
> 
> Great effort. 🎉

I agree with Ahmed that inheritance isn't the best solution here so I will create another PR and we can discuss it in conversation @ahmdmhasn @Ahattalla

File: Healthy/Classes/ReusableViews/UILoadingButton/AnimatableView.swift (hunk 5 above)
Comment (ahmdmhasn) [reply Healthy/Classes/ReusableViews/UILoadingButton/AnimatableView.swift]: ❤️

File: Healthy/Classes/Extensions/Collection+Helpers.swift [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/Collection+Helpers.swift]:
```suggestion
```

File: Healthy/Classes/ReusableViews/AnimatableView/AnimatableView.swift [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/AnimatableView/AnimatableView.swift]:
```suggestion
```

File: Healthy/Classes/ReusableViews/AnimatableView/AnimatableView.swift [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+protocol AnimatableView: AnyObject {
```
Comment (ahmdmhasn) [inline Healthy/Classes/ReusableViews/AnimatableView/AnimatableView.swift]:
```suggestion
protocol AnimatableView {
```

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import UIKit
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift:1]:
```swift
extension UIButton: AnimatableView {

    // MARK: AnimatableView Conformance
    
    func startAnimating() {
        titleLabel?.alpha = 0
        isEnabled = false

        let activityIndicator = self.activityIndicator ?? createAndCenterActivityIndicator()
        activityIndicator.startAnimating()
    }

    func stopAnimating() {
        titleLabel?.alpha = 1
        isEnabled = true

        activityIndicator?.stopAnimating()
    }

    // MARK: Private Handlers

    private var activityIndicator: UIActivityIndicatorView? {
        subviews
            .filter { $0 is UIActivityIndicatorView }
            .first as? UIActivityIndicatorView
    }
    
    private func createAndCenterActivityIndicator() -> UIActivityIndicatorView {
        let activityIndicator = createActivityIndicator()
        addSubview(activityIndicator)
        centerActivityIndicatorInButton(activityIndicator)
        return activityIndicator
    }
    
    private func createActivityIndicator() -> UIActivityIndicatorView {
        let activityIndicator = UIActivityIndicatorView()
        activityIndicator.hidesWhenStopped = true
        activityIndicator.color = .white
        return activityIndicator
    }

    private func centerActivityIndicatorInButton(_ activityIndicator: UIView) {
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false

        NSLayoutConstraint.activate([
            activityIndicator.centerXAnchor.constraint(equalTo: centerXAnchor),
            activityIndicator.centerYAnchor.constraint(equalTo: centerYAnchor)
        ])
    }
}
```

**Review by ahmdmhasn — APPROVED** [review]

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            addSubview(activityIndicator)
+            centerActivityIndicatorInButton()
+        }
+        showSpinning()
+    }
+
+    func stopAnimating() {
+        isEnabled = true
+        setTitleColor(titleLabel?.textColor.withAlphaComponent(0.5), for: .disabled)
+
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift:30]:
```suggestion
        // ...
        setTitleColor(titleLabel?.textColor.withAlphaComponent(0.5), for: .disabled)
```

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        return activityIndicator
+    }
+
+    private func centerActivityIndicatorInButton() {
+        guard let activityIndicator else {
+            assertionFailure("Trying to configure constraints for invalid activity indicator.")
+            return
+        }
+
+        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift]:
```suggestion
    private func centerActivityIndicatorInButton(_ activityIndicator: UIView) {
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
```

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        NSLayoutConstraint.activate([
+            activityIndicator.centerXAnchor.constraint(equalTo: centerXAnchor),
+            activityIndicator.centerYAnchor.constraint(equalTo: centerYAnchor)
+        ])
+    }
+
+    private func showSpinning() {
+        guard let activityIndicator else { return }
+        activityIndicator.startAnimating()
+        activityIndicator.isHidden = false
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift]:
```suggestion
```

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        isEnabled = false
+        setTitleColor(.clear, for: .disabled)
+
+        titleLabel?.alpha = 0
+        if activityIndicator == nil {
+            let activityIndicator = createActivityIndicator()
+            addSubview(activityIndicator)
+            centerActivityIndicatorInButton()
+        }
+        showSpinning()
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift]:
```suggestion
        
        activityIndicator?.startAnimating()
```

File: Healthy/Classes/Extensions/UIButton+AnimatableView.swift [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    func stopAnimating() {
+        isEnabled = true
+        setTitleColor(titleLabel?.textColor.withAlphaComponent(0.5), for: .disabled)
+
+        titleLabel?.alpha = 1
+
+        guard let activityIndicator else { return }
+        activityIndicator.stopAnimating()
+        activityIndicator.removeFromSuperview()
```
Comment (ahmdmhasn) [inline Healthy/Classes/Extensions/UIButton+AnimatableView.swift]:
```suggestion
        activityIndicator?.stopAnimating()
        activityIndicator?.removeFromSuperview()
```

**Review by ahmdmhasn — APPROVED** [review]
🎉🎉🎉🚀🚀🚀

**Review by a7maad-ayman — APPROVED** [review]

**Pull request merged**